### ISSUE TITLE
PATCH implementation's "to SCIM, from SCIM" behaviour is problematic

### ISSUE THREAD

**Issue by pond** [RIPAGlobal/scimitar#6]
Scimitar has a mostly-working PATCH implementation which does a full to-SCIM, patch-that-Hash, from-SCIM cycle to read-modify-write data into a mapped resource. This was the best I could get my brain to figure out at the time.

Problems arise if a to-SCIM cycle on a resource does not contain all data, since the from-SCIM operation uses PUT / replace-like semantics intentionally, and if there's anything missing, it looks like a set-to-`nil` operation. This was recently revealed in a User mapping in an application using Scimitar, which supports Roles. Users can set up mappings between role names in the client system and role names in the Scimitar-based service providing application, and only mapped role data was being shown in the SCIM representation (since unmapped service provider roles would mean nothing to the client). When PATCHing e.g. just the 'is_active' field of a User, the result would be that the Roles collection got modified too, because any unmapped roles were omitted in the to-SCIM operation and the subsequent PUT-like from-SCIM operation treated this as a rewrite-roles-array-contents action.

I can't quite figure out how to make it work still and it'll take some time, but the right thing to do would be to traverse the to-SCIM data as we do today (required in order for filters in the path to work) _at the same time_ as traversing the corresponding underlying class's attribute map. This is difficult for filters in particular - completely different filter application schemes will be needed for matching data, versus matching an attribute array static map, or an attribute array dynamic map. If that can be figured out, though, by the time a given PATCH path has finished traversal, we'll know both the SCIM attribute to be added/removed/altered and have attribute map data to allow that operation to be applied immediately via write accessors - no from-SCIM step would be required.

As a bonus, it'd be good if we could track the no-filter attribute map path arising so that the resource schema can be consulted and mutability flags read for a given attribute, helping to trap invalid PATCH attempts to in-context immutable attribute data that might otherwise be permitted.

**Comment by pelted** [comment]
I believe I'm running into these issues from [Azure AD SCIM Validator](https://scimvalidator.microsoft.com/). Looking at some ways to resolve this when I came across this issue.

**Comment by gsar** [comment]
@pond @kuldeepaggarwal @pelted FWIW, i'm working around the problem with this rather ugly workaround. i don't think we should expect `userName` to be sent in anything except `create`.

``` ruby
module Scim
  class UsersController < Scimitar::ActiveRecordBackedResourcesController
    def update
      id = params[:id]
      # work around failure of PATCH endpoints in Microsoft Entra SCIM validator
      # see https://github.com/RIPAGlobal/scimitar/issues/6#issuecomment-1645714989
      unless params[:userName]
        resource = storage_scope.find_by(id: id)
        if resource
          params[:userName] = resource.scim_user_name
        end
      end
      super
    end
    ...
  end
end
```

**Comment by pond** [comment]
I've started looking at this, in relation to #80 -> #109 where it came up as part of under-hood issues. I note the SCIM RFC on PUT has some interesting wording:

https://datatracker.ietf.org/doc/html/rfc7644#section-3.5.1

> Attributes whose mutability is "readWrite" that are omitted from
> the request body MAY be assumed to be not asserted by the client.
> The service provider MAY assume that any existing values are to be
> cleared, or the service provider MAY assign a default value to the
> final resource representation.  Service providers MAY take into
> account whether or not a client has access to, or understands, all
> of the resource's attributes when deciding whether non-asserted
> attributes SHALL be removed or defaulted.  Clients that want to
> override a server's defaults MAY specify "null" for a
> single-valued attribute, or an empty array "[]" for a multi-valued
> attribute, to clear all values.

So ignoring for a moment the stretch goal of properly obeying read-only property mutability, the existing implementation's from-SCIM-to-SCIM behaviour becomes less problematic if the to-SCIM part uses the "MAY" clauses above and chooses to treat an unspecified attribute as _leave alone_ rather than the current _flatten all_ behaviour. If `PUT` / replace works via "leave untouched attributes alone", it becomes easier to have the `PATCH` behaviour keep track of things it changes and then run a PUT-style splat of changed-only attributes on top of the existing resource.

While I have some reservations about making a behavioural change that _might_ be breaking, the set-to-nil behaviour seems to be causing more trouble than not. I can't imagine many cases where it wouldn't simply look like a bug. That said, I really do hate it when software upon which I rely breaks behaviour (even for good reasons) and my code fails without warning, so maybe I shouldn't Be That Guy!

**Comment by pond** [comment]
I _think_ #109 does enough work to essentially close this issue. It's not the prettiest approach but it ended up being less code than a full-on 100% attribute-map-based solution.

**Comment by pond** [comment]
Fixed in [v2.7.2](https://rubygems.org/gems/scimitar/versions/2.7.2) / [v1.8.2](https://rubygems.org/gems/scimitar/versions/1.8.2).
